src/NuGet.Core/NuGet.Packaging/Signing/Signatures/PrimarySignature.cs

```diff
@@ -0,0 +1,164 @@
+            {
+                ms.CopyTo(stream);
+            }
+        }
+
+        /// <summary>
+        /// Retrieve the bytes of the signed cms signature.
+        /// </summary>
+        public byte[] GetBytes()
+        {
+            return SignedCms.Encode();
+        }
+
+        /// <summary>
+        /// Create a signature based on a valid signed cms
+        /// </summary>
+        /// <param name="cms">signature data</param>
+        public static PrimarySignature Load(SignedCms cms)
+        {
+            if (cms == null)
+            {
+                throw new ArgumentNullException(nameof(cms));
+            }
+
+            if (cms.SignerInfos.Count != 1)
+            {
+                throw new SignatureException(NuGetLogCode.NU3009, Strings.Error_NotOnePrimarySignature);
+            }
+
+            var signerInfo = cms.SignerInfos[0];
+            VerifySigningCertificate(cms, signerInfo, SigningSpecifications.V1);
+
+            return PrimarySignatureFactory.CreateSignature(cms);
+        }
+
+        /// <summary>
+        /// Create a signature based on a valid byte array to be decoded as a signed cms
+        /// </summary>
+        /// <param name="data">signature data</param>
+        public static PrimarySignature Load(byte[] data)
+        {
+            if (data == null)
+            {
+                throw new ArgumentNullException(nameof(data));
+            }
+
+            var cms = new SignedCms();
+            cms.Decode(data);
+
+            return Load(cms);
+        }
+
+        /// <summary>
+        /// Create a signature based on a valid byte stream to be decoded as a signed cms
+        /// </summary>
+        /// <param name="stream">signature data</param>
+        public static PrimarySignature Load(Stream stream)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            using (stream)
+            using (var ms = new MemoryStream())
+            {
+                stream.CopyTo(ms);
+                return Load(ms.ToArray());
+            }
+        }
+
+        public override byte[] GetSignatureHashValue(HashAlgorithmName hashAlgorithm)
+        {
+            using (var authorSignatureNativeCms = NativeCms.Decode(SignedCms.Encode(), detached: false))
+            {
+                return NativeCms.GetSignatureValueHash(hashAlgorithm, authorSignatureNativeCms);
+            }
+        }
+
+        protected PrimarySignature(SignedCms signedCms, SignatureType signatureType)
+            : base(GetSignerInfo(signedCms), signatureType)
+        {
+            SignedCms = signedCms;
+            SignatureContent = SignatureContent.Load(SignedCms.ContentInfo.Content, SigningSpecifications.V1);
+        }
+
+        protected override void ThrowForInvalidSignature()
+        {
+            ThrowForInvalidPrimarySignature();
+        }
+
+        protected static void ThrowForInvalidPrimarySignature()
```

Comment:
Not sure this method is needed.  Just move the body into `ThrowForInvalidSignature()`.

Reply:
I decided to have this method not as part of `ThrowForInvalidSignature()` because it is usefull as a static method in some of the static methods in this class, and because `ThrowForInvalidSignature()` is an abstract method it cannot be static. Therefore I decided to have another method that is static with the desired implementation and just let `ThrowForInvalidSignature()` call this method.